src/Chainweb/Pact/PactService.hs

```diff
@@ -477,97 +480,135 @@ finalizeCheckpointer finalize = do
 _liftCPErr :: Either String a -> PactServiceM cas a
 _liftCPErr = either internalError' return
 
-type Balances = HM.HashMap Text Decimal
 
 data Uniqueness = Duplicate | Unique
 
+-- | Performs a dry run of PactExecution's `buyGas` function for transactions being validated.
+--
+attemptBuyGas
+    :: PayloadCas cas
+    => Checkpointer
+    -> Miner
+    -> Vector (ChainwebTransaction, Uniqueness)
+    -> PactServiceM cas (Vector (ChainwebTransaction, Uniqueness, Bool))
+attemptBuyGas cp miner txs = withDiscardedBatch $ do
+  mbLatestBlock <- liftIO $ _cpGetLatestBlock cp
+  (bh, bhash) <- case mbLatestBlock of
+                     Nothing -> throwM NoBlockValidatedYet
+                     (Just !p) -> return p
+  let target = Just (succ bh, bhash)
+  withCheckpointer target "validateChainwebTxs/attemptBuyGas" $ \(PactDbEnv' dbEnv) -> do
+      psEnv <- ask
+      res <- V.fromList . ($ []) . sfst <$> V.foldM (f psEnv dbEnv) (T2 id mempty) txs
+      return $! Discard res
+  
+  where
+    f psEnv dbEnv (T2 dl mcache) cmd = do
+      T2 !res mcache' <- runBuyGas psEnv dbEnv mcache cmd
+      pure $! T2 (dl . (res :)) mcache'
+
+    createGasEnv
+        :: PayloadCas cas
+        => PactServiceEnv cas
+        -> P.PactDbEnv a
+        -> P.Command (P.Payload P.PublicMeta P.ParsedCode)
+        -> P.CommandEnv a
+    createGasEnv envM db cmd =
+      P.CommandEnv Nothing P.Transactional db logger
+      freeGasEnv publicData spv nid
+      where
+        !logger = _cpeLogger . _psCheckpointEnv $ envM
+        !publicData = set P.pdPublicMeta (publicMetaOf cmd) (_psPublicData envM)
+        !spv = _psSpvSupport envM
+        !nid = networkIdOf cmd
+
+    runBuyGas
+        :: PayloadCas cas
+        => PactServiceEnv cas
+        -> P.PactDbEnv a
+        -> ModuleCache
+        -> (ChainwebTransaction, Uniqueness)
+        -> PactServiceM cas (T2 (ChainwebTransaction, Uniqueness, Bool) ModuleCache)
+    runBuyGas _ _ mcache (tx, u@Duplicate) = return $ T2 (tx, u, False) mcache
+    runBuyGas envM db mcache (tx, u@Unique) = do
```

Comment:
`Duplicate` and `Unique` are singletons, so we can match directly on them instead of binding them to a variable. GHC will inline this appopriately.